File: _submissions/18UEE071_Ayush/Day4/countingTriangles.py
class Solution:
    # @param A : list of integers
    # @return an integer
    def getNextGreater(self, arr, x, start):
        
        end = len(arr) - 1
        ans = -1

        while(start<=end):
            mid = start + (end-start)//2

            if(arr[mid] < x):
                start = mid+1
                ans = mid
            else:
                end = mid-1
        return ans

    def nTriang(self, arr):
        arr = sorted(arr)
        count = 0
        mod = 1000000007
        for i in range(len(arr)-2):
            for j in range(i+1, len(arr)-1):
                sumOfTwoSides = arr[i] + arr[j]
                greaterSide = self.getNextGreater(arr, sumOfTwoSides, j+1)
                # print(greaterSide)
                if(greaterSide == -1):
                    continue
                count+=(greaterSide - j )
        return count%mod


arr = [1, 1, 1, 2, 2]
ob = Solution()
print(ob.nTriang(arr))

File: _submissions/19UCS159_Tharun/Day4/countingTriangles.java
import java.util.*;

public class countingTriangles {
  public int nTriang(int[] A) {
    Arrays.sort(A);
    int n = A.length;
    long ans = 0, mod = (long) 1e9 + 7;
    for (int i = 0; i < n - 2; i++) {
      for (int j = i + 1; j < n - 1; j++) {
        int sum = A[i] + A[j];
        int k = findNextGreater(A, j + 1, sum);
        // System.out.println(i+" "+j+" "+k);
        ans = (ans + (k - j - 1)) % mod;
      }
    }
    return (int) ans;
  }

  private int findNextGreater(int[] A, int start, int sum) {
    int l = start, r = A.length - 1;
    int ans = A.length;
    while (l <= r) {
      int mid = l + (r - l) / 2;
      if (sum <= A[mid]) {
        ans = mid;
        r = mid - 1;
      } else {
        l = mid + 1;
      }
    }
    return ans;
  }

}
